src/MassTransit/Pipeline/Filters/ProcessingTimeoutFilter.cs

```diff
@@ -0,0 +1,49 @@
+﻿namespace MassTransit.Pipeline.Filters
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using GreenPipes;
+
+
+    public class ProcessingTimeoutFilter<TContext, TResult> :
+        IFilter<TContext>
+        where TContext : class, ConsumeContext
+        where TResult : TContext
+    {
+        readonly Func<TContext, CancellationToken, TResult> _contextFactory;
+        readonly TimeSpan _timeout;
+
+        public ProcessingTimeoutFilter(Func<TContext, CancellationToken, TResult> contextFactory, TimeSpan timeout)
+        {
+            _contextFactory = contextFactory;
+            _timeout = timeout;
+        }
+
+        public async Task Send(TContext context, IPipe<TContext> next)
+        {
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken);
+
+            cts.CancelAfter(_timeout);
+
+            try
+            {
+                var todoRenameContext = _contextFactory(context, cts.Token);

```

Comment:
TODO

Reply:
Fixed alongside few extra TODOs in comments.